File: pkg/utils/zfs.go
// SPDX-License-Identifier: BSD-2-Clause
//
// Copyright (c) 2025 The FreeBSD Foundation.
//
// This software was developed by Hayzam Sherif <hayzam@alchemilla.io>
// of Alchemilla Ventures Pvt. Ltd. <hello@alchemilla.io>,
// under sponsorship from the FreeBSD Foundation.

package utils

import (
	"strings"
	zfsServiceInterfaces "sylve/internal/interfaces/services/zfs"
)

func ParseZpoolListOutput(pools string, vdevs string) (*zfsServiceInterfaces.Zpool, error) {
	poolSlice := strings.Split(strings.TrimSpace(pools), "\n")
	vdevSlice := strings.Split(strings.TrimSpace(vdevs), "\n")

	zpool := &zfsServiceInterfaces.Zpool{}

	for _, pool := range poolSlice {
		if pool == "" {
			continue
		}

		parts := strings.Fields(pool)
		if len(parts) < 9 {
			continue
		}

		zpool.Name = parts[0]
		zpool.Health = parts[1]
		zpool.Allocated = StringToUint64(parts[2])
		zpool.Size = StringToUint64(parts[3])
		zpool.Free = StringToUint64(parts[4])
		zpool.ReadOnly = parts[5] == "on"
		zpool.Freeing = StringToUint64(parts[6])
		zpool.Leaked = StringToUint64(parts[7])
		zpool.DedupRatio = StringToFloat64(parts[8])

		for _, vdev := range vdevSlice {
			if strings.HasPrefix(vdev, zpool.Name) {
				continue
			}

			vdevParts := strings.Fields(vdev)

			if len(vdevParts) < 7 {
				continue
			}

			vdev := zfsServiceInterfaces.Vdev{}

			vdev.Name = vdevParts[0]
			vdev.Alloc = StringToUint64(vdevParts[1])
			vdev.Free = StringToUint64(vdevParts[2])
			vdev.Operations.Read = StringToUint64(vdevParts[3])
			vdev.Operations.Write = StringToUint64(vdevParts[4])
			vdev.Bandwidth.Read = StringToUint64(vdevParts[5])
			vdev.Bandwidth.Write = StringToUint64(vdevParts[6])

			zpool.Vdevs = append(zpool.Vdevs, vdev)
		}
	}

	return zpool, nil
}

File: web/src/lib/utils/zfs.ts
/**
 * SPDX-License-Identifier: BSD-2-Clause
 *
 * Copyright (c) 2025 The FreeBSD Foundation.
 *
 * This software was developed by Hayzam Sherif <hayzam@alchemilla.io>
 * of Alchemilla Ventures Pvt. Ltd. <hello@alchemilla.io>,
 * under sponsorship from the FreeBSD Foundation.
 */

import type { Zpool } from '$lib/types/zfs/pool';

export function isDeviceVdev(device: string, pools: Zpool[]): boolean {
	if (pools.length === 0) {
		return false;
	}

	for (const pool of pools) {
		for (const vdev of pool.vdevs) {
			if (vdev.name === device || vdev.name === `/dev/${device}`) {
				return true;
			}
		}
	}

	return false;
}

export function getHealthHelpers(health: string): { icon: string; color: string; text: string } {
	switch (health) {
		case 'ONLINE':
			return {
				icon: 'carbon:checkmark-filled',
				color: 'text-green-600 dark:text-green-500',
				text: 'Online'
			};
		case 'DEGRADED':
			return {
				icon: 'carbon:warning-filled',
				color: 'text-yellow-600 dark:text-yellow-500',
				text: 'Degraded'
			};
		case 'FAULTED':
			return {
				icon: 'carbon:close-filled',
				color: 'text-red-600 dark:text-red-500',
				text: 'Faulted'
			};
		case 'OFFLINE':
			return {
				icon: 'material-symbols:offline-pin-off',
				color: 'text-red-600 dark:text-red-500',
				text: 'Offline'
			};
		case 'UNAVAIL':
			return {
				icon: 'carbon:warning-alt-filled',
				color: 'text-yellow-600 dark:text-yellow-500',
				text: 'Unavailable'
			};
		case 'REMOVED':
			return {
				icon: 'carbon:warning-alt-filled',
				color: 'text-yellow-600 dark:text-yellow-500',
				text: 'Removed'
			};
		default:
			return {
				icon: 'carbon:warning-alt-filled',
				color: 'text-yellow-600 dark:text-yellow-500',
				text: 'Unknown'
			};
	}
}

export function isValidPoolName(name: string): boolean {
	const reserved = ['log', 'mirror', 'raidz', 'raidz1', 'raidz2', 'raidz3', 'spare'];

	if (!name) return false;
	if (reserved.some((r) => name.startsWith(r))) return false;
	if (!/^[a-zA-Z]/.test(name)) return false;
	if (!/^[a-zA-Z0-9_.-]+$/.test(name)) return false;
	if (name.includes('%')) return false;
	if (/^c[0-9]/.test(name)) return false;

	return true;
}

export function isValidDatasetName(name: string): boolean {
	if (!name || typeof name !== 'string') return false;
	if (name.length > 255) return false;
	if (/[^\x21-\x7E]/.test(name)) return false;
	if (name.includes('%') || name.includes(' ')) return false;

	const components = name.split('/');
	for (const comp of components) {
		if (!comp) return false;
		if (!/^[a-zA-Z0-9_.-]+$/.test(comp)) return false;
		if (comp.startsWith('.') || comp.startsWith('-')) return false;
	}

	return true;
}
